File: app/src/main/java/com/reteno/sample/SampleApp.java
package com.reteno.sample;


import android.app.Application;
import android.os.Handler;

import androidx.annotation.NonNull;

import com.reteno.core.Reteno;
import com.reteno.core.RetenoApplication;
import com.reteno.core.RetenoConfig;
import com.reteno.core.RetenoImpl;
import com.reteno.core.domain.model.event.LifecycleTrackingOptions;
import com.reteno.core.identification.DeviceIdProvider;
import com.reteno.core.lifecycle.ScreenTrackingConfig;
import com.reteno.sample.util.AppSharedPreferencesManager;

import java.util.ArrayList;

public class SampleApp extends Application implements RetenoApplication {

    private Reteno retenoInstance;

    @Override
    public void onCreate() {
        super.onCreate();
        if (AppSharedPreferencesManager.getShouldDelayLaunch(this)) {
            AppSharedPreferencesManager.setDelayLaunch(this, false);
            retenoInstance = new RetenoImpl(this);
            Handler handler = new Handler();
            handler.postDelayed(() -> {
                retenoInstance.initWith(new RetenoConfig(false, createProvider(), LifecycleTrackingOptions.Companion.getALL(), BuildConfig.API_ACCESS_KEY));
            }, 3000L);
        } else {
            retenoInstance = new RetenoImpl(this, BuildConfig.API_ACCESS_KEY, new RetenoConfig(false, createProvider()));
        }
        ArrayList<String> excludeScreensFromTracking = new ArrayList<String>();
        excludeScreensFromTracking.add("NavHostFragment");
        retenoInstance.autoScreenTracking(new ScreenTrackingConfig(false, excludeScreensFromTracking));
    }

    private DeviceIdProvider createProvider() {
        DeviceIdProvider provider = null;
        int deviceIdDelay = AppSharedPreferencesManager.getDeviceIdDelay(this);
        String deviceId = AppSharedPreferencesManager.getDeviceId(this);
        if (!deviceId.isEmpty()) {
            long startTime = System.currentTimeMillis();
            provider = () -> {
                if (System.currentTimeMillis() - startTime > deviceIdDelay) {
                    return deviceId;
                } else {
                    return null;
                }
            };
        }
        return provider;
    }


    @NonNull
    @Override
    public Reteno getRetenoInstance() {
        return retenoInstance;
    }
}

File: app/src/main/java/com/reteno/sample/SampleApp.kt
package com.reteno.sample

import android.app.Application
import android.widget.Toast
import androidx.work.Configuration
import com.reteno.core.Reteno
import com.reteno.core.RetenoConfig
import com.reteno.core.identification.DeviceIdProvider
import com.reteno.core.lifecycle.ScreenTrackingConfig
import com.reteno.core.util.toStringVerbose
import com.reteno.push.RetenoNotifications
import com.reteno.sample.util.AppSharedPreferencesManager.getDeviceId
import com.reteno.sample.util.AppSharedPreferencesManager.getDeviceIdDelay
import com.reteno.sample.util.AppSharedPreferencesManager.getShouldDelayLaunch

class SampleApp : Application(), Configuration.Provider {

    override fun getWorkManagerConfiguration(): Configuration {
        return Configuration.Builder()
            .setMinimumLoggingLevel(android.util.Log.INFO)
            .build()
    }

    override fun onCreate() {
        super.onCreate()
        if (!getShouldDelayLaunch(this)) {
            Reteno.initWithConfig(
                RetenoConfig.Builder()
                    .pauseInAppMessages(false)
                    .customDeviceIdProvider(createProvider())
                    .accessKey(BuildConfig.API_ACCESS_KEY)
                    .setDebug(BuildConfig.DEBUG)
                    .build()
            )
        }
        val excludeScreensFromTracking = ArrayList<String>()
        excludeScreensFromTracking.add("NavHostFragment")
        Reteno.instance.autoScreenTracking(ScreenTrackingConfig(false, excludeScreensFromTracking))
        RetenoNotifications.click.addListener {
            val text = "Push click: ${it.toStringVerbose()}"
            Toast.makeText(this, text, Toast.LENGTH_SHORT)
                .show()
        }
        RetenoNotifications.custom.addListener {
            val text = "Custom push received: ${it.toStringVerbose()}"
            Toast.makeText(this, text, Toast.LENGTH_SHORT)
                .show()
        }
        RetenoNotifications.close.addListener {
            val text = "Push closed: ${it.toStringVerbose()}"
            Toast.makeText(this, text, Toast.LENGTH_SHORT)
                .show()
        }
        RetenoNotifications.received.addListener {
            val text = "Push received: ${it.toStringVerbose()}"
            Toast.makeText(this, text, Toast.LENGTH_SHORT)
                .show()
        }
        RetenoNotifications.inAppCustomDataReceived.addListener {
            val text = "InApp custom data received: $it"
            Toast.makeText(this, text, Toast.LENGTH_SHORT)
                .show()
        }
    }

    fun createProvider(): DeviceIdProvider? {
        var provider: DeviceIdProvider? = null
        val deviceIdDelay = getDeviceIdDelay(this)
        val deviceId = getDeviceId(this)
        if (deviceId!!.isNotEmpty()) {
            val startTime = System.currentTimeMillis()
            provider = DeviceIdProvider {
                if (System.currentTimeMillis() - startTime > deviceIdDelay) {
                    return@DeviceIdProvider deviceId
                } else {
                    return@DeviceIdProvider null
                }
            }
        }
        return provider
    }
}
